### PR TITLE
Image Upload: update navigation UX

### PR DESCRIPTION
**Pull request by jaclync**
Navigation UX update for #1713 

## Goals

This PR updates the following navigation UX (ref: p99K0U-2mN-p2#comment-4732):

- **(Done)** If the user taps “Done” while some images are still being uploaded, we will keep track of the images that are in-progress, display the in-progress UI (blurred image + circle spinner) for these images on the Edit Product screen, and then update the UI for these images once they are uploaded
  - **(This PR)** When the user taps “Update” on the Edit Product screen, we first show the in-progress UI, wait for all images to be uploaded and then update the Product remotely.
  - **(This PR)** When the user navigates away from the Edit Product screen, no changes are made to the Product and these images keep being uploaded in the background.
- **(This PR)** If the user navigates away from the Edit Product Images screen while some images are still being uploaded, these images keep being uploaded in the background but they won’t be added to the Product.

## Changes

- Updated UX on the Edit Product screen when the user taps "Update" while some images are still being uploaded:
  - In `ProductFormViewController`, added a product images observation block to wait for all product images to be uploaded before we update the product remotely
- Updated UX on the Edit Product screen when the user navigates away while some images are still being uploaded:
  - In `ProductFormViewController`, created a private helper `func hasUnsavedChanges(product: Product) -> Bool` to check for any unsaved changes from the latest product and show the "Update" CTA and discard changes action sheet when the user navigates away
- Updated UX on the Edit Product Images screen when the user navigates away:
  - Added `resetProductImages` to `ProductImageActionHandler` for resetting the changes to product images + unit test case
  - In `ProductImagesViewController`, showed discard changes action sheet when there are any local images, and reset images to the product from init when the user decides to navigate away
- Added an extension helper `hasImagesPendingUpload ` to `ProductImageStatus` collection, and  renamed `ProductImageStatus+ImagesTests` to `ProductImageStatus+HelpersTests` with new test cases on `hasImagesPendingUpload`

## Testing

Notes:
- please test this on a simulator/device below iOS 13.4! otherwise the back navigation callback wouldn't work for testing the discard changes action sheet behavior (https://github.com/woocommerce/woocommerce-ios/issues/2089)
- if your internet is too fast, you might want to try slowing down the internet (like with the Developer settings > Network Link Conditioner on a device) so that an image is being uploaded during the following testing steps 😆 

### Case 1: tap "Done" on the Product Images screen while some images are still being uploaded, then tap "Update" on the Edit Product screen

- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell in the images header
- Tap "Add Photos"
- Choose a photo or take one with camera --> there should be a spinner cell after picking a photo **(note: please make sure to perform the steps below while this image is being uploaded)**
- While the image is being uploaded, quickly tap "Done" in the navigation bar --> it should go back to the Edit Product screen, the image that is still being uploaded has a spinner
- Quickly tap "Update" --> the fullscreen in-progress should be shown, and when it completes all the images should be uploaded and attached to the product
- Check the Media Library and the product in wp-admin --> the image should be uploaded and the product has the new image

### Case 2: tap "Done" on the Product Images screen while some images are still being uploaded, then navigate away from the Edit Product screen

- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell in the images header
- Tap "Add Photos"
- Choose a photo or take one with camera --> there should be a spinner cell after picking a photo **(note: please make sure to perform the steps below while this image is being uploaded)**
- While the image is being uploaded, quickly tap "Done" in the navigation bar --> it should go back to the Edit Product screen, the image that is still being uploaded has a spinner
- Quickly tap "<" in the navigation bar --> a discard changes action should be shown
- Quickly tap "Discard changes" to confirm --> it should leave the product screen
- After a few seconds, check the Media Library and the product in wp-admin --> the image should be uploaded and the product doesn't have the new image

### Case 3: navigate away on the Product Images screen while some images are still being uploaded

- Go to the Products tab
- Tap on a simple Product
- Tap on the "+" cell in the images header
- Tap "Add Photos"
- Choose a photo or take one with camera --> there should be a spinner cell after picking a photo **(note: please make sure to perform the steps below while this image is being uploaded)**
- While the image is being uploaded, quickly tap "<" in the navigation bar --> a discard changes action should show
- Quickly tap "Discard changes" to confirm --> it should land back on the edit product screen without the image picked/taken previously, and the "Update" CTA shouldn't be visible in the navigation bar
- After a few seconds, check the Media Library and the product in wp-admin --> the image should be uploaded and the product doesn't have the new image

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/Media/ProductImageStatus.swift [hunk 1]
```diff
@@ -24,6 +24,19 @@ extension Collection where Element == ProductImageStatus {
             }
         }
     }
+
+    /// Whether there are still any images being uploaded.
+    ///
+    var hasImagesPendingUpload: Bool {
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Products/Media/ProductImageStatus.swift:30]: I believe there's enough context to just say `hasPendingUpload`. What do you think? 🙂

File: WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageStatus+HelpersTests.swift [hunk 2]
```diff
@@ -0,0 +1,50 @@
+import Photos
+import XCTest
+
+@testable import WooCommerce
+@testable import Yosemite
+
+final class ProductImageStatus_HelpersTests: XCTestCase {
+    // MARK: - `images`
+
+    func testWithoutRemoteImages() {
```
[the remark this answers is not in the document]
Comment (shiki) [reply WooCommerce/WooCommerceTests/ViewRelated/Products/Media/ProductImageStatus+HelpersTests.swift:10]: Oh, I see that this is not new code. Nevermind!

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift [hunk 3]
```diff
@@ -166,6 +167,30 @@ private extension ProductFormViewController {
...
+
+            self.product = self.productUpdater.imagesUpdated(images: productImageStatuses.images)
+            group.leave()
+        }
+
+        group.notify(queue: .main) { [weak self] in
+            observationToken.cancel()
+            self?.updateProductRemotely()
+        }
+    }
```
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:191]: ⚠️  I feel like this **and** `updateProductRemotely` should be in the `ViewModel`. For now, how do you feel about placing this inside a separate function like:

```swift
func waitUntilAllImagesAreUploaded(_ completion: () -> ()) {

}

func dispatchUpdateProductAction() {
}

func uploadProductRemotely() {
    waitUntilAllImagesAreUploaded {
        dispatchUpdateProductAction()
    }
}
```

Or something like that. 😄 

----

This is nice clean code by the way! I like that `group.notify()` shows what will happen _next_.

**Comment by jaclync** [comment]
> Case 2
I wasn't able to replicate it. When I discard changes, I'm not able to see the image uploaded on wp-admin. I tried several times, also without the network link conditioner, but I wasn't able. Can you try on my website? Maybe there is a problem on my side.
> Case 3
Same as case 2, since I'm not able to upload the image when I left the product detail screen.

thanks for testing @pmusolino ! I just tested these 2 cases with your store, and the images appeared after about 2 minutes 😆 (if the app is closed before the upload completes, the upload is lost. but if the app remains open or even backgrounded, I believe it's still going to be uploaded eventually unless the connection drops)

Screen Shot 2020-04-14 at 4 34 38 PM

> Case 1 ✅
I just noticed while I was testing the first test case, that if the connection is really slow (SDL or Edge in the Network link conditioner) the image upload takes a lot of time, which is ok, but when I press the update button of the product I don't have any visual feedback for minutes, and this makes me think that loading has stopped. I also get stuck on this screen, and the only way I can get out of it is to close the app (iOS 12.2). It's a UX problem, not a code problem, but I think that having a progress bar here it's not easy.

this is the screen you were stuck on right? yea I agree there's not much visual feedback and it's blocking any other UX, not sure if it's worth adding a "Cancel" CTA on this screen to allow the user to cancel the product update? it'd also be nice to show a progress bar like you suggested (like from the number of images that are pending upload, plus the product update - the logic could get complicated but feasible) so it doesn't appear stuck. @Garance91540 do you think we could add a "Cancel" CTA and/or show a progress bar on this waiting screen?

**Comment by pmusolino** [comment]
Thanks Jaclyn for verifying my issue! I'll retry shortly, but I never put the app in the background.

> this is the screen you were stuck on right?

Yes correct 👍 

> not sure if it's worth adding a "Cancel" CTA on this screen to allow the user to cancel the product update?

I was testing the app on iOS 12.2, where you can't dismiss the view (because the modal presentation style is full screen). It might be worth the effort adding the cancellation action even to the dismiss of the view?

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift (hunk 3 above)
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift]: yea the logic in `ProductFormViewController` is certainly getting out of hand and these user interactions handling logic could be in the view model! updated the function names in https://github.com/woocommerce/woocommerce-ios/pull/2126/commits/964615ce07f56dd6cb2b3d79d80ebcdff0acbc75 for now.  I'm planning on taking on https://github.com/woocommerce/woocommerce-ios/issues/1985 toward the end of Products M2 or even sooner after this PR

**Comment by jaclync** [comment]
@shiki thanks for the suggestions! I updated the PR based on your comments for another look 🙂

File: WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift (hunk 3 above)
Comment (shiki) [reply WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift]: Fine by me! We don't need to refactor all the time and not hit product goals. 🙂 Thanks for making the change!

**Comment by jaclync** [comment]
> I'll retry shortly, but I never put the app in the background.

Putting the app in the background should be fine, as long as the connection isn't lost (my device was on Edge with Network Link Conditioner). @pmusolino I'm merging this now, but lemme know if you still see issues with image upload after leaving the edit product screens when you get a chance to test this!

**Comment by pmusolino** [comment]
@jaclync I was able to test it, and now it's working. Maybe I get some networking issues switching between the network link conditioner and my default network? 🤷 I don't know, but now it's working properly.

**Comment by Garance91540** [comment]
I can't test on my device, but the logic you outlined @jaclync seems 👌

Tagging @nbradbury  here since I believe he's working on implementing the same logic on Android

**Pull request merged**